File: app/src/main/java/com/arnaudpiroelle/manga/model/Page.java
package com.arnaudpiroelle.manga.model;

import android.os.Parcel;
import android.os.Parcelable;

public class Page implements Parcelable{
    private String mangaAlias;
    private String chapterNumber;
    private String pageNumber;
    private String pageName;
    private String extension;

    public Page() {

    }

    private Page(Parcel in) {
        mangaAlias = in.readString();
        chapterNumber = in.readString();
        pageNumber = in.readString();
        pageName = in.readString();
        extension = in.readString();
    }

    public String getMangaAlias() {
        return mangaAlias;
    }

    public void setMangaAlias(String mangaAlias) {
        this.mangaAlias = mangaAlias;
    }

    public String getChapterNumber() {
        return chapterNumber;
    }

    public void setChapterNumber(String chapterNumber) {
        this.chapterNumber = chapterNumber;
    }

    public String getPageNumber() {
        return pageNumber;
    }

    public void setPageNumber(String pageNumber) {
        this.pageNumber = pageNumber;
    }

    public String getExtension() {
        return extension;
    }

    public void setExtension(String extension) {
        this.extension = extension;
    }

    public String getPageName() {
        return pageName;
    }

    public void setPageName(String pageName) {
        this.pageName = pageName;
    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeString(mangaAlias);
        dest.writeString(chapterNumber);
        dest.writeString(pageNumber);
        dest.writeString(pageName);
        dest.writeString(extension);
    }

    public static final Parcelable.Creator<Page> CREATOR = new Parcelable.Creator<Page>() {
        public Page createFromParcel(Parcel in) {
            return new Page(in);
        }

        public Page[] newArray(int size) {
            return new Page[size];
        }
    };

    public static class PageBuilder {
        private Page page;

        public PageBuilder() {
            page = new Page();
        }

        public static PageBuilder createPage() {
            return new PageBuilder();
        }

        public PageBuilder withPageNumber(String pageNumber) {
            page.setPageNumber(pageNumber);
            return this;
        }

        public PageBuilder withMangaAlias(String mangaAlias) {
            page.setMangaAlias(mangaAlias);
            return this;
        }

        public PageBuilder withChapterNumber(String chapterNumber) {
            page.setChapterNumber(chapterNumber);
            return this;
        }

        public PageBuilder withExtenstion(String extension) {
            page.setExtension(extension);
            return this;
        }

        public PageBuilder withName(String name) {
            page.setPageName(name);
            return this;
        }

        public Page build() {
            return page;
        }
    }
}

File: app/src/main/kotlin/com/arnaudpiroelle/manga/model/Page.kt
package com.arnaudpiroelle.manga.model

import android.os.Parcel
import android.os.Parcelable

class Page : Parcelable {
    var mangaAlias: String? = null
    var chapterNumber: String? = null
    var pageNumber: String? = null
    var pageName: String? = null
    var extension: String? = null

    constructor() {

    }

    private constructor(`in`: Parcel) {
        mangaAlias = `in`.readString()
        chapterNumber = `in`.readString()
        pageNumber = `in`.readString()
        pageName = `in`.readString()
        extension = `in`.readString()
    }

    override fun describeContents(): Int {
        return 0
    }

    override fun writeToParcel(dest: Parcel, flags: Int) {
        dest.writeString(mangaAlias)
        dest.writeString(chapterNumber)
        dest.writeString(pageNumber)
        dest.writeString(pageName)
        dest.writeString(extension)
    }

    public class PageBuilder {
        private var page: Page = Page()

        public fun withPageNumber(pageNumber: String): PageBuilder {
            page.pageNumber = pageNumber
            return this
        }

        public fun withMangaAlias(mangaAlias: String): PageBuilder {
            page.mangaAlias = mangaAlias
            return this
        }

        public fun withChapterNumber(chapterNumber: String): PageBuilder {
            page.chapterNumber = chapterNumber
            return this
        }

        public fun withExtenstion(extension: String): PageBuilder {
            page.extension = extension
            return this
        }

        public fun withName(name: String): PageBuilder {
            page.pageName = name
            return this
        }

        public fun build(): Page {
            return page;
        }

        companion object {
            public fun createPage(): PageBuilder {
                return PageBuilder();
            }
        }
    }

    companion object {

        val CREATOR: Parcelable.Creator<Page> = object : Parcelable.Creator<Page> {
            override fun createFromParcel(`in`: Parcel): Page {
                return Page(`in`)
            }

            override fun newArray(size: Int): Array<Page?> {
                return arrayOfNulls(size)
            }
        }
    }
}
